### ISSUE TITLE
Wgpu crashes with particular glsl code during verification.

### ISSUE THREAD

**Issue by iMplode-nZ** [gfx-rs/wgpu#1197]
**Description**
Wgpu throws an error during verification of this following glsl code:

```glsl
#version 450

layout(location = 0) in ivec3 a_position;
layout(location = 1) in ivec2 a_entity_voxel_id;

layout(location = 0) out vec4 v_color;
layout(location = 1) out mat3 v_rotation;
layout(location = 4) out vec3 v_position;

layout(set = 0, binding = 0)
uniform Uniforms {
    mat4 u_window_to_clip_space_matrix;
    mat4 u_projection_matrix_window_space;
    vec3 u_position;
    vec2 u_window_size;
};

layout(set = 2, binding = 0) readonly buffer BEntityRotations {
    mat3 b_entity_rotation[];
};
layout(set = 2, binding = 1) readonly buffer BEntityPositions {
    vec4 b_entity_position[];
};
layout(set = 2, binding = 2) readonly buffer BTypeColors {
    vec4 b_type_color[];
};

void add_voxel_corner_to_aabb(
    mat4 projection_matrix_window_space, vec2 window_size,
    vec3 position, float half_size, mat3 rotation,
    inout vec2 aabb_min, inout vec2 aabb_max,
    vec3 corner_id
) {
    vec3 corner = position + rotation * (half_size * corner_id);
    vec4 screen_pos_h = (u_projection_matrix_window_space * vec4(corner, 1));
    vec2 screen_pos = screen_pos_h.xy / screen_pos_h.w;
    screen_pos = clamp(screen_pos, vec2(0), window_size);
    aabb_min = min(aabb_min, screen_pos);
    aabb_max = max(aabb_max, screen_pos);
}

void main() {
    v_color = b_type_color[a_entity_voxel_id.y];
    v_rotation = b_entity_rotation[a_entity_voxel_id.x];
    v_position = b_entity_position[a_entity_voxel_id.x].xyz + v_rotation * a_position - u_position;
    vec2 aabb_min = u_window_size;
    vec2 aabb_max = vec2(0);
    #define PROJ(x, y, z) add_voxel_corner_to_aabb(u_projection_matrix_window_space, u_window_size, v_position, 1.0 / 2.0, v_rotation,\
        aabb_min, aabb_max, vec3(x, y, z))
    PROJ(-1,-1,-1);
    PROJ(-1,-1, 1);
    PROJ(-1, 1,-1);
    PROJ(-1, 1, 1);
    PROJ( 1,-1,-1);
    PROJ( 1,-1, 1);
    PROJ( 1, 1,-1);
    PROJ( 1, 1, 1);
    aabb_min = floor(aabb_min);
    aabb_max = ceil(aabb_max);
    vec2 aabb_size = aabb_max - aabb_min;
    gl_PointSize = max(aabb_size.x, aabb_size.y) + 2;
    gl_Position = u_window_to_clip_space_matrix * vec4((aabb_max + aabb_min) / 2, 1.0, 1.0);
    gl_Position.z = gl_Position.w;
}

```

The error message is:

```
wgpu error: Validation Error

Caused by:
    In Device::create_shader_module
      note: label = `shader.vert.spv`
    Entry point main at Vertex is invalid: Function(Resolve(FunctionReturnsVoid))
```

If all of the calls to `add_voxel_corner_to_aabb` are removed, the error does not happen.
The error does not happen in the earlier version of the code, which replaced the buffers with textures.

**Repro steps**

I think it should be enough to use this shader.vert code in a wgpu application.

**Platform**
Ubuntu 20.04, wgpu 0.7.

**Comment by kvark** [comment]
Thank you for the report! We'll fix this for sure.
In the meantime, you can avoid passing the `VALIDATION` flag to shader module creation.

**Comment by iMplode-nZ** [comment]
Why does this only happen when I have buffers though? I had an earlier version of the code where the buffers were replaced with textures, and that worked fine/

**Comment by kvark** [comment]
This shouldn't have anything to do with buffers. It has to do with `add_voxel_corner_to_aabb` calls, which is a void function. You can probably work around it by adding return type to this function, temporarily.
Are you sure the buffer-related validation error is the same?

**Comment by iMplode-nZ** [comment]
I had this issue only after I did this change to my code: https://github.com/Counterproduction-game/Counterproduction/compare/8f4215845b9c14dbc052f8c565a54a4dd84344cb...develop#diff-5203e5cc0ae6c65d825edda2883cf87c34d2d7a5526b2ee2b2245feec468eab9, so I think it does have stuff to do with the buffers.

**Comment by kvark** [comment]
Are you sure the error is exactly the one you posted, coming from this vertex shader? Could you, perhaps, attach the SPIR-V zip here, and I'll be happy to run Naga over it locally.

**Comment by iMplode-nZ** [comment]
Spirv zip? The compiler I use (shaderc) generates .spv files...

**Comment by kvark** [comment]
Github doesn't allow dropping SPV, but it allows dropping TXT and ZIP. So if you ZIP your shader, you can drop it here.

**Comment by iMplode-nZ** [comment]
@kvark 
[shader.zip](https://github.com/gfx-rs/wgpu/files/5938130/shader.zip)

**Comment by kvark** [comment]
Thank you! So this shader shader used to fail validation, and now it validates in Naga.
What about the version of the shader that was no failing the validation, but also used this void function?

**Comment by iMplode-nZ** [comment]
https://github.com/Counterproduction-game/Counterproduction/blob/8f4215845b9c14dbc052f8c565a54a4dd84344cb/render/src/shader.vert here is the earlier code.

**Comment by kvark** [comment]
Ok I converted it to SPIR-V and ran revision f563b5810bac6458809899c5563af7c89140ece5 on it.
First, it failed validation on `spirv::Capability::Sampled1D` capability, then it (after working around it) failed on `Function(Resolve(FunctionReturnsVoid))`.

So what happened is - when you only had buffers, it was actually validating and failing. But when you replaced buffers with textures, we failed to parse SPIR-V, and thus we couldn't validate it at all.

Let's close this issue then as solved, and I'm going to follow-up with a few more caps enabled in SPIR-V frontend.
